planetiler-core/src/main/java/com/onthegomap/planetiler/pmtiles/WriteablePmtiles.java

```diff
@@ -0,0 +1,619 @@
+
+  public record Directories(byte[] root, byte[] leaves, int numLeaves) {}
+
+  private static Directories buildRootLeaves(List<Entry> subEntries, int leafSize) throws IOException {
+    ArrayList<Entry> rootEntries = new ArrayList<>();
+    ByteArrayList leavesOutputStream = new ByteArrayList();
+    int leavesLength = 0;
+    int numLeaves = 0;
+
+    for (int i = 0; i < subEntries.size(); i += leafSize) {
+      numLeaves++;
+      int end = i + leafSize;
+      if (i + leafSize > subEntries.size()) {
+        end = subEntries.size();
+      }
+      byte[] leafBytes = serializeDirectory(subEntries, i, end);
+      leafBytes = Gzip.gzip(leafBytes);
+      rootEntries.add(new Entry(subEntries.get(i).tileId, leavesLength, leafBytes.length, 0));
+      leavesOutputStream.add(leafBytes);
+      leavesLength += leafBytes.length;
+    }
+
+    byte[] rootBytes = serializeDirectory(rootEntries, 0, rootEntries.size());
+    rootBytes = Gzip.gzip(rootBytes);
+
+    return new Directories(rootBytes, leavesOutputStream.toArray(), numLeaves);
+  }
+
+  /**
+   * Serialize all entries into bytes, choosing the # of leaf directories to ensure the header+root fits in 16 KB.
+   *
+   * @param entries a sorted ObjectArrayList of all entries in the tileset.
+   * @return byte arrays of the root and all leaf directories, and the # of leaves.
+   * @throws IOException
+   */
+  public static Directories makeRootLeaves(List<Entry> entries) throws IOException {
+    if (entries.size() < 16384) { // coarse heuristic to short circuit
+      byte[] testBytes = serializeDirectory(entries, 0, entries.size());
+      testBytes = Gzip.gzip(testBytes);
+
+      if (testBytes.length < INIT_SECTION - HEADER_LEN) {
+        return new Directories(testBytes, new byte[0], 0);
+      }
+    }
+
+    int leafSize = (int) Math.max(entries.size() / 3_500d, 4096);
+
+    while (true) {
+      Directories temp = buildRootLeaves(entries, leafSize);
+      if (temp.root.length < INIT_SECTION - HEADER_LEN) {
+        return temp;
+      }
+      leafSize *= 1.2;
+    }
+  }
+
+  private WriteablePmtiles(SeekableByteChannel channel) throws IOException {
+    this.out = channel;
+    out.write(ByteBuffer.wrap(new byte[INIT_SECTION]));
+  }
+
+  public static WriteablePmtiles newWriteToFile(Path path) throws IOException {
+    return new WriteablePmtiles(
+      FileChannel.open(path, StandardOpenOption.CREATE, StandardOpenOption.READ, StandardOpenOption.WRITE));
+  }
+
+  public static WriteablePmtiles newWriteToMemory(SeekableInMemoryByteChannel bytes) throws IOException {
+    return new WriteablePmtiles(bytes);
+  }
+
+  @Override
+  public TileOrder tileOrder() {
+    return TileOrder.HILBERT;
+  }
+
+  @Override
+  public void initialize(PlanetilerConfig config, TileArchiveMetadata tileArchiveMetadata, LayerStats layerStats) {
+    this.layerStats = layerStats;
+    this.tileArchiveMetadata = tileArchiveMetadata;
+  }
+
+  @Override
+  public void finish(PlanetilerConfig config) {
```

Comment:
I might be interested in seeing some logs for some of these steps, like sorting, each round of the adaptive leaf sizing, then writing to the end/beginning.  We don't need super detailed timing - just a log saying the step has started will capture the timestamp to give us a sense for where the time goes.

Reply:
added more logging to `finish`